File: go/domain/cart_test.go
// ABOUTME: Unit tests for the Cart domain type.
// ABOUTME: Verifies cart creation, line item management, and subtotal calculation.
package domain

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestNewCart(t *testing.T) {
	cart := NewCart()
	assert.Empty(t, cart.Lines)
}

func TestCartAddLine(t *testing.T) {
	cart := NewCart()

	product := Product{
		Sku:      MustSku("APPLE"),
		Name:     "Apple",
		Category: "Fruit",
	}
	line := LineItem{
		Product:   product,
		Quantity:  MustQuantity(3),
		UnitPrice: Euros(1.50),
	}

	cart.AddLine(line)

	assert.Len(t, cart.Lines, 1)
	assert.Equal(t, line, cart.Lines[0])
}

func TestCartSubtotal(t *testing.T) {
	tests := []struct {
		name     string
		lines    []LineItem
		expected Money
	}{
		{
			name:     "empty cart",
			lines:    []LineItem{},
			expected: Euros(0.00),
		},
		{
			name: "single line",
			lines: []LineItem{
				{
					Product: Product{
						Sku:      MustSku("APPLE"),
						Name:     "Apple",
						Category: "Fruit",
					},
					Quantity:  MustQuantity(3),
					UnitPrice: Euros(1.50),
				},
			},
			expected: Euros(4.50),
		},
		{
			name: "multiple lines",
			lines: []LineItem{
				{
					Product: Product{
						Sku:      MustSku("APPLE"),
						Name:     "Apple",
						Category: "Fruit",
					},
					Quantity:  MustQuantity(3),
					UnitPrice: Euros(1.50),
				},
				{
					Product: Product{
						Sku:      MustSku("BANANA"),
						Name:     "Banana",
						Category: "Fruit",
					},
					Quantity:  MustQuantity(2),
					UnitPrice: Euros(0.75),
				},
			},
			expected: Euros(6.00),
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cart := Cart{Lines: tt.lines}
			subtotal := cart.Subtotal()
			assert.Equal(t, tt.expected, subtotal)
		})
	}
}

File: typescript/src/domain/Cart.test.ts
import { describe, it, expect } from 'vitest';
import { Cart } from './Cart';
import { LineItem } from './LineItem';
import { Product } from './Product';
import { Quantity } from './Quantity';
import { Money } from './Money';
import { Sku } from './Sku';

describe('Cart', () => {
  describe('construction', () => {
    it('should construct with empty lines', () => {
      const cart = new Cart([]);
      expect(cart.lines).toHaveLength(0);
    });

    it('should construct with line items', () => {
      const product = new Product(new Sku('SKU-A'), 'Product A', 'Electronics');
      const lineItem = new LineItem(product, new Quantity(1), Money.euros('10.00'));
      const cart = new Cart([lineItem]);

      expect(cart.lines).toHaveLength(1);
      expect(cart.lines[0]).toBe(lineItem);
    });
  });

  describe('subtotal', () => {
    it('should return zero for empty cart', () => {
      const cart = new Cart([]);
      const subtotal = cart.subtotal();
      expect(subtotal.amount.toFixed(2)).toBe('0.00');
    });

    it('should calculate subtotal for single line', () => {
      const product = new Product(new Sku('SKU-A'), 'Product A', 'Electronics');
      const lineItem = new LineItem(product, new Quantity(2), Money.euros('10.00'));
      const cart = new Cart([lineItem]);

      const subtotal = cart.subtotal();
      expect(subtotal.amount.toFixed(2)).toBe('20.00');
    });

    it('should calculate subtotal for multiple lines', () => {
      const productA = new Product(new Sku('SKU-A'), 'Product A', 'Electronics');
      const productB = new Product(new Sku('SKU-B'), 'Product B', 'Books');

      const lineItemA = new LineItem(productA, new Quantity(2), Money.euros('10.00'));
      const lineItemB = new LineItem(productB, new Quantity(1), Money.euros('25.00'));
      const cart = new Cart([lineItemA, lineItemB]);

      const subtotal = cart.subtotal();
      // 2 * 10 + 1 * 25 = 45
      expect(subtotal.amount.toFixed(2)).toBe('45.00');
    });

    it('should handle rounding across multiple lines', () => {
      const productA = new Product(new Sku('SKU-A'), 'Product A', 'Electronics');
      const productB = new Product(new Sku('SKU-B'), 'Product B', 'Books');

      const lineItemA = new LineItem(productA, new Quantity(3), Money.euros('1.235'));
      const lineItemB = new LineItem(productB, new Quantity(2), Money.euros('2.225'));

      const cart = new Cart([lineItemA, lineItemB]);

      const subtotal = cart.subtotal();
      // Line A: 1.24 * 3 = 3.72
      // Line B: 2.23 * 2 = 4.46
      // Total: 3.72 + 4.46 = 8.18
      expect(subtotal.amount.toFixed(2)).toBe('8.18');
    });
  });

  describe('immutability', () => {
    it('should not allow external modification of lines', () => {
      const product = new Product(new Sku('SKU-A'), 'Product A', 'Electronics');
      const lineItem = new LineItem(product, new Quantity(1), Money.euros('10.00'));
      const cart = new Cart([lineItem]);

      // TypeScript should prevent this at compile time,
      // but we test runtime immutability
      expect(() => {
        // @ts-expect-error - Testing runtime immutability
        cart.lines.push(lineItem);
      }).toThrow();
    });
  });
});
